### ISSUE TITLE
Request for a way to indicate how a class/concepts fits into "layers" or "areas of interest"

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter#6974]
As we learn from UX studies, we noticed that we don't currently have an easy way for a developer to understand in which "area" they are looking at a concept. For example, if they are looking at the Padding class in the API docs, "where" are they? What kind of conceptual thing is this?

(Note: our library organization in the framework is too coarse-grained to convey this grouping. For example, there are a lot of conceptual areas inside the widget library.)

Vision: a user looking at a thing in the API docs should easily understand in which group/AreaOfInterest/Layer it belongs, and then easily see other related things. For example, when looking at Padding, the user can easily see "This is part of the _layout_ area of interest. Oh, these are the other helpful layout ideas."

Strawman: a way to annotate the thing (class/function/etc) to put it in a "doc grouping". Then, the dartdoc tool could be taught how to indicate which group the thing is in.

**Comment by sethladd** [comment]
cc @Hixie

**Comment by Hixie** [comment]
Proposal:

Only top-level identifiers have grouping (libraries, classes, enums, typedefs, constants, top-level functions, top-level properties). They can have multiple groups.

An identifier with an explicit `@DocGroup('')` annotation gains that string as a group.

An identifier in a library gains the groping of that library.

A subclass gains the groupings from its superclass (via `extends`, not `implements` or `with`).

**Comment by Hixie** [comment]
So then `library widgets` would get `@DocGroup('Widgets Layer')`, the `Widget` class would get `@DocGroup('Widgets')`, and the `Padding` class would get `@DocGroup('Layout Widgets')`.

**Comment by Hixie** [comment]
The grouping would have hierarchy:
- top-level group from `library` annotation
- annotations from superclasses, from `Object` down to the class itself
- inner-most group from own annotations

So the `Padding` example above would be:

 - Widgets Layer
    - Widgets
       - Layout Widgets
          - `Padding`

**Comment by jcollins-g** [comment]
This is dart-lang/dartdoc#1353

**Comment by goderbauer** [comment]
The upstream dartdoc issue has been resolved a long time ago. Is there anything left to do here in the flutter/flutter repo?

**Comment by jcollins-g** [comment]
Yes, there is something to do still.  Dartdoc now supports organizing via categories, but the flutter repo needs to have the proper tags in the source code to make that happen.   Since that's extremely labor-intensive to do manually, probably some kind of migration script needs to translate categories.json to the appropriate tags in the source code.  Either that, or we need to add a new bug to dartdoc to load the categories in from the manually created json file (just have dartdoc populate category information internally as though it had been read from source code).  This latter path might be desirable as the categories.json file may contain or wish to contain fields that are irrelevant from a source code perspective.

**Comment by InMatrix** [comment]
@mjohnsullivan @emshack @brianegan Would you consider addressing this issue as part of your ongoing improvements to Flutter's API docs? A proposal is in @jcollins-g's comment above.
